index.bs

```diff
@@ -130,6 +130,71 @@ much like a data property as possible. Specific guidance in this regard includes
     true. (This will not always be the case, e.g. if a normalization or type conversion step is
     necessary, but should be held as a goal for normal code paths.)
 
+<h3 id="live-vs-static">Consider whether objects should be live or static</h3>
+
+Objects returned from functions, attribute getters, etc.,
+can either be live or static.
+A <dfn>live object</dfn> is one that continues to reflect
+changes made after it was returned to the caller.
+A <dfn>static object</dfn> is one that reflects
+the state at the time it was returned.
+
+Objects that are the way state is mutated are generally live.
+For example, DOM nodes are returned as live objects,
+since they are the API through which attributes are set and other changes
+to the DOM are made.
+They also reflect changes to the DOM made in other ways
+(such as through user interaction with forms).
+Even though the nodes themselves are live,
+they might be returned from a static object,
+such as a static {{NodeList}} that is
+the result of {{ParentNode/querySelectorAll()}}.
+
+Objects that represent a collection that might change over time
+should generally be returned as static objects.
+This is because it is confusing to users of the API when
+a collection changes while being iterated.
+Because of this,
+it is generally considered a mistake that methods like
+{{Document/getElementsByTagName()}} return live objects;
+{{ParentNode/querySelectorAll()}} was made to return static objects
+as a result of this experience.
+
+Note:  It's possible that some of this advice should be reconsidered
+for <a>Iterable</a>, <a>Maplike</a>, and <a>Setlike</a> types
+which can be iterated using iterators that have
+well-defined behavior for mutation that happens during iteration.
+(Is this true for all three?)
```

Comment:
Er, I probably should have said reasonable rather than well-defined.  @bzbarsky was talking yesterday about how iteration of ES maps is defined by describing the map as an array (in the spec, not in implementations) so that things added get added to the end so that in-progress iteration will find them, and so that elements that are removed leave conceptual "holes" so as not to mess up in-progress iteration.  It wasn't 100% clear to me whether all of this definition applies to maplike as well (as opposed to actual maps).

But if there's a widely expected behavior for mutation while iterating that isn't completely crazy and doesn't require developers to figure out how to adjust an index (or to make sure to iterate from length-1 to 0 and only touch elements already reached), it seems like it might be a little more reasonable to expose that behavior, particularly if (as @bzbarsky pointed out yesterday) iteration is possible *only* through iterators and not through indexed getters.